sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/WarnUpgradedDependencies.hs

```diff
@@ -0,0 +1,121 @@
+module DA.Daml.LF.TypeChecker.WarnUpgradedDependencies
+  ( checkModule
+  ) where
+
+import           Control.Monad.Extra
+import           Data.Foldable (for_)
+
+import DA.Daml.LF.Ast
+import DA.Daml.LF.TypeChecker.Env
+import DA.Daml.LF.TypeChecker.Error
+import Data.Functor.Foldable (cataA, embed)
+import DA.Daml.LF.Ast.Recursive (TypeF(..))
+
+-- | Check whether a type refers to 
+checkType :: forall m. MonadGamma m => Bool -> Type -> m ()
+checkType isSerializable typ = do
+    _ <- cataA go typ
+    pure ()
+  where
+    go :: TypeF (m Type) -> m Type
+    go typeF = do
+      case typeF of
+        TConF tcn -> checkTypeCon isSerializable tcn
+        _ -> pure ()
+      embed <$> sequence typeF
+
+checkTypeCon :: (MonadGamma m) => Bool -> Qualified TypeConName -> m ()
+checkTypeCon isSerializable name = do
+  case qualPackage name of
+    PRSelf -> pure ()
+    PRImport pkgId -> do
+      pkg <- inWorld (lookupPackage (PRImport pkgId))
+
+      -- When using a datatype from a new (>= LF1.17) daml-script in any type
+      -- definition
+      case packageMetadata pkg of
+        Nothing -> pure ()
+        Just meta@PackageMetadata { packageName } ->
+          let isNewDamlScript :: Bool
+              isNewDamlScript = packageName `elem` map PackageName ["daml-script-lts", "daml-script-lts-stable"]
+          in
+          when isNewDamlScript $ do
+            diagnosticWithContext $ WEDependsOnDatatypeFromNewDamlScript (pkgId, meta) (packageLfVersion pkg) name
+
+      -- When using an LF1.15 serializable datatype in a serializable position while
+      -- targeting >= LF1.17
+      when isSerializable $ do
+        targetLfVersion <- getLfVersion
+        datatype <- inWorld (lookupDataType name)
+        when (packageLfVersion pkg == version1_15 && targetLfVersion `supports` featurePackageUpgrades && getIsSerializable (dataSerializable datatype)) $ do
+          diagnosticWithContext $ WEUpgradeDependsOnSerializableNonUpgradeableDataType (pkgId, packageMetadata pkg, packageLfVersion pkg) targetLfVersion name
+
+-- | Check whether a data type definition satisfies all serializability constraints.
+checkDataType :: MonadGamma m => DefDataType -> m ()
+checkDataType dataType = do
+  let isSerializable = getIsSerializable (dataSerializable dataType)
+  case dataCons dataType of
+    DataEnum {} -> pure ()
+    DataInterface {} -> pure ()
+    DataVariant constructors -> forM_ constructors (checkType isSerializable . snd)
+    DataRecord fields -> forM_ fields (checkType isSerializable . snd)
+  pure ()
+
+-- | Check whether a template satisfies all serializability constraints.
+checkTemplate :: MonadGamma m => Module -> Template -> m ()
+checkTemplate mod0 tpl = do
+  let tcon = Qualified PRSelf (moduleName mod0) (tplTypeCon tpl)
+  checkType True (TCon tcon)
+  for_ (tplChoices tpl) $ \ch -> withContext (ContextTemplate mod0 tpl $ TPChoice ch) $ do
+    checkType True (snd (chcArgBinder ch))
```

Comment:
NB. currently the argbinder check doesn't do anything because it's always a TCon from the local.